csharp/src/Apache.Arrow/Arrays/MapArray.cs

```diff
@@ -0,0 +1,174 @@
+                ValidityBufferBuilder.Resize(length + 1);
+                return this;
+            }
+
+            public Builder Clear()
+            {
+                ValueOffsetsBufferBuilder.Clear();
+                KeyBuilder.Clear();
+                ValueBuilder.Clear();
+                ValidityBufferBuilder.Clear();
+                return this;
+            }
+
+        }
+
+        public StructArray KeyValues => base.Values as StructArray;
+        public IArrowArray Keys => KeyValues.Fields[0];
+        public new IArrowArray Values => KeyValues.Fields[1];
+
+        public MapArray(IArrowType dataType, int length,
+            ArrowBuffer valueOffsetsBuffer, IArrowArray structs,
+            ArrowBuffer nullBitmapBuffer, int nullCount = 0, int offset = 0)
+            : this(
+                  new ArrayData(
+                      dataType, length, nullCount, offset, new[] { nullBitmapBuffer, valueOffsetsBuffer },
+                      new[] { structs.Data }
+                  ), structs)
+        {
+        }
+
+        public MapArray(ArrayData data)
+            : this(data, ArrowArrayFactory.BuildArray(data.Children[0]))
+        {
+        }
+
+        private MapArray(ArrayData data, IArrowArray structs) : base(data, structs, ArrowTypeId.Map)
+        {
+        }
+
+        public IEnumerable<Tuple<K, V>> GetTuples<TKeyArray, K, TValueArray, V>(int index, Func<TKeyArray, int, K> getKey, Func<TValueArray, int, V> getValue)
+            where TKeyArray : Array where TValueArray : Array
+        {
+            ReadOnlySpan<int> offsets = ValueOffsets;
+            // Get key values
+            int start = offsets[index];
+            int end = offsets[index + 1];
+            StructArray array = KeyValues.Slice(start, end - start) as StructArray;
+
+            TKeyArray keyArray = array.Fields[0] as TKeyArray;
+            TValueArray valueArray = array.Fields[1] as TValueArray;
+
+            for (int i = start; i < end; i++)
+            {
+                yield return new Tuple<K, V>(getKey(keyArray, i), getValue(valueArray, i));
+            }
+        }
+
+        public IEnumerable<KeyValuePair<K,V>> GetKeyValuePairs<TKeyArray, K, TValueArray, V>(int index, Func<TKeyArray, int, K> getKey, Func<TValueArray, int, V> getValue)
+            where TKeyArray : Array where TValueArray : Array
+        {
+            ReadOnlySpan<int> offsets = ValueOffsets;
+            // Get key values
+            int start = offsets[index];
+            int end = offsets[index + 1];
+            StructArray array = KeyValues.Slice(start, end - start) as StructArray;
+
+            TKeyArray keyArray = array.Fields[0] as TKeyArray;
+            TValueArray valueArray = array.Fields[1] as TValueArray;
+
+            for (int i = start; i < end; i++)
+            {
+                yield return new KeyValuePair<K,V>(getKey(keyArray, i), getValue(valueArray, i));
+            }
+        }
```

Comment:
These APIs are a little more complex than others (due to the need for `getKey` and `getValue`.  It shouldn't be done as part of this PR but I wonder if we might want to start working on "examples" or a "cookbook" for C# at some point.  Accessing map values would be a good example.

Reply:
Ideally I would have preferred use something like c++ scalars, but it would need to be an abstract method Array.GetScalar(index) usable by all array data types, where we wouldn't need these external methods

Reply:
I can also imagine that there's a more-neutral scale-friendly base implementation and some opinionated helpers for making it easier to work with the base. But yeah, some samples are definitely a good idea.